pipeline/main.py

```diff
@@ -1,279 +0,0 @@
-import torch
-import torch.optim as optim
-import torch.nn as nn
-import torch.nn.functional as F
-import numpy as np
-import dgl
-import dgl.function as fn
-
-from dgl.data import CoraGraphDataset, CiteseerGraphDataset, PubmedGraphDataset
-from tqdm import trange
-
-class MixHopConv(nn.Module):
-    """
-    One layer of MixHopGCN.
-    """
-    def __init__(self,
-                 in_dim,
-                 out_dim,
-                 p=[0, 1, 2],
-                 dropout=0,
-                 activation=None,
-                 batchnorm=False):
-        super(MixHopConv, self).__init__()
-        self.in_dim = in_dim
-        self.out_dim = out_dim
-        self.p = p
-        self.activation = activation
-        self.batchnorm = batchnorm
-
-        # define dropout layer
-        self.dropout = nn.Dropout(dropout)
-
-        # define batch norm layer
-        if self.batchnorm:
-            self.bn = nn.BatchNorm1d(out_dim * len(p))
-        
-        # define weight dict for each power j
-        self.weights = nn.ModuleDict({
-            str(j): nn.Linear(in_dim, out_dim) for j in p
-        })
-
-    def forward(self, graph, feats):
-        with graph.local_scope():
-            degs = graph.in_degrees().float().clamp(min=1)
-            norm = torch.pow(degs, -0.5).to(feats.device).unsqueeze(1)
-            max_j = max(self.p) + 1
-            outputs = []
-            for j in range(max_j):
-                feats = feats * norm
-                graph.ndata['h'] = feats
-                graph.update_all(fn.copy_u('h', 'm'), fn.sum('m', 'h'))
-                feats = graph.ndata.pop('h')
-                feats = feats * norm
-
-                if j in self.p:
-                    output = self.weights[str(j)](feats)
-                    outputs.append(output)
-            
-            final = torch.cat(outputs, dim=1)
-            
-            if self.batchnorm:
-                final = self.bn(final)
-            
-            if self.activation is not None:
-                final = self.activation(final)
-            
-            final = self.dropout(final)
-
-            return final
-
-class MixHopGCN(nn.Module):
-    def __init__(self,
-                 in_dim,
-                 hid_dim, 
-                 out_dim,
-                 num_layers=2,
-                 p=[0, 1, 2],
-                 input_dropout=0.0,
-                 layer_dropout=0.0,
-                 activation=None,
-                 batchnorm=False):
-        super(MixHopGCN, self).__init__()
-        self.in_dim = in_dim
-        self.hid_dim = hid_dim
-        self.out_dim = out_dim
-        self.num_layers = num_layers
-        self.p = p
-        self.input_dropout = input_dropout
-        self.layer_dropout = layer_dropout
-        self.activation = activation
-        self.batchnorm = batchnorm
-
-        self.layers = nn.ModuleList()
-
-        # Input layer and initial q for output layer
-        # self.n_segs = int(self.hid_dim / self.out_dim)
```

Comment:
I thought they only said "divide s_l columns into sets of size c"?

Reply:
See [here](https://github.com/samihaija/mixhop/blob/master/mixhop_model.py), line 17 ~ 30.

Reply:
ditto